etc/format.py

```diff
@@ -0,0 +1,272 @@
+
+    parent_dir = fpath.parent
+
+    def f(mat: re.Match[str]) -> str:
+        # See groups in INCLUDE_RE
+        target = mat['path']
+        abs_target = parent_dir / target
+        if abs_target.is_file():
+            # This should be a relative include:
+            newl = f'{mat["directive"]}"./{target}"{mat["tail"]}'
+        else:
+            newl = f'{mat["directive"]}<{target}>{mat["tail"]}'
+        print(f' - {fpath}: update #include directive: {mat[0]!r} → {newl!r}', file=sys.stderr)
+        return newl
+
+    return f
+
+
+def _fixup_includes(fpath: Path, *, mode: RunMode) -> bool:
+    """
+    Apply #include-fixup to the content of the given source file.
+    """
+    # Split into lines
+    old_lines = fpath.read_text().split('\n')
+    # Do a regex substitution on ever line:
+    rx = re.compile(INCLUDE_RE, re.VERBOSE)
+    new_lines = [rx.sub(_include_subst_fn(fpath), ln) for ln in old_lines]
+    # Did we change anything?
+    did_change = new_lines != old_lines
+    # We changed something. What do we do next?
+    match did_change, mode:
+        case False, _:
+            # No file changes. Nothing to do
+            return True
+        case _, 'apply':
+            # We are applying changes. Write the lines back into the file and tell
+            # the caller that we succeeded
+            fpath.write_text('\n'.join(new_lines), newline='\n')
+            return True
+        case _, 'check':
+            # File changes, and we are only checking. Print an error message and indicate failure to the caller
+            print(f'File [{fpath}] contains improper #include directives', file=sys.stderr)
+            return False
+
+
+def _format_files(files: Iterable[Path], *, mode: RunMode, clang_format: str) -> bool:
+    """
+    Run clang-format on some files, and fixup the #includes in those files.
+    """
+
+    def fixup_one(p: Path) -> bool:
+        try:
+            return _fixup_includes(p, mode=mode)
+        except Exception as e:
+            raise RuntimeError(f'Unexpected error while fixing-up the #includes on file [{p}] (See above)') from e
+
+    # First update the `#include` directives, since that can change the sort order
+    # that clang-format might want to apply
+    if not all(list(map(fixup_one, files))):
+        return False
+
+    # Whether we check for format violations or modify the files in-place
+    match mode:
+        case 'apply':
+            mode_args = ['-i']
+        case 'check':
+            mode_args = ['--dry-run', '-Werror']
+    cmd = [clang_format, *mode_args, *map(str, files)]
+    try:
+        res = subprocess.run(cmd, check=False, stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
+    except Exception as e:
+        raise RuntimeError(f'Failed to spawn [{clang_format}] process for formatting files (See above)') from e
+    sys.stderr.buffer.write(res.stdout)
+    return res.returncode == 0
+
+
+def _get_files_matching(pat: str) -> Sequence[Path]:
```

Comment:
This function is not used (it's not used in the C repo either, I just haven't gotten around to deleting it yet)

Reply:
Posted https://github.com/mongodb/mongo-c-driver/pull/2102.